### ISSUE TITLE
Subsurface scattering

### ISSUE THREAD

**Issue by GameSMRT** [GameSMRT/Lead-To-Hell-Art-Build#8]
Discussing Subsurface scattering

**Comment by justinjacot** [comment]
Sooo tough to find an ideal solution, because we want a shader that also handles thin materials, like paper.  HDRP's SSS actually works really well and does all we need it to, but the fidelity is low, just like the volumetrics.  Hopefully these two features will be more usable when the final release hits

**Comment by GameSMRT** [comment]
I was able to get the Lux SSS shader options working again finally.
https://assetstore.unity.com/packages/vfx/shaders/lux-physically-based-shader-framework-16000
https://github.com/larsbertram69/Lux-2.02-Personal

The package has not been updated itself to work with Unity 2018. But forum posts have instructions to do so.

Here is a video of some tests. Unfortunately, there seems to be a lot of compression artifacts on the video. Best I could do for now.
Our reference image is on the left. In the middle is a 3d paper crane, on the right is part of a crane made out of single-sided geo.
I added some post effects to the scene, but no other enhancements, so the shadows don't look too good.
https://youtu.be/5U1aLOdtSbk - Both cranes have the same shader settings
https://youtu.be/AJVXvseiCKg - The single-sided crane has it's scattering power set to the setting they intend for cards

I need to do more testing with perhaps combining it with translucency and generating a SSS map to get the folded look of paper, rather than just a solid color.

**Comment by justinjacot** [comment]
It looks good so far. If this scene is available in the art build, we can start experimenting with our models to see how it looks. We can combine it with one my lighting templates. It’s one of the last missing pieces to the visual stack

**Comment by justinjacot** [comment]
Can you put this scene up?  I wanna mess with it and integrate it

**Comment by GameSMRT** [comment]
Not letting you have all the fun :P (watch in full 1080p)
https://youtu.be/94pt1NZuUy0

Pretty damn close now.
Now with Aura, NGSS, Your Cinematic AA. The grain was causing the compression artifacts so I removed it for now.

**Comment by justinjacot** [comment]
Send me that scene!  I wanna use the shader and mess with it!!

**Comment by justinjacot** [comment]
That looks sick though man...its pretty much there

**Comment by GameSMRT** [comment]
Unfortunately I don't think we can use this shader. Everytime I start Unity, the SSS on the shader is broken like you saw in person. I have to Delete the Libaray folder in the project to force Unity to reimport everything to fix it. It may not work if the game is built.

I am going to look at other SSS shaders that are actually compatible with 2018. Preferably one that isn't an uber shader with alot or features we aren't going to use.

**Comment by justinjacot** [comment]
Ok cool.  We just need to make sure it can do double-sided materials since this is paper after all.  The one that looked best that I looked into was this
https://assetstore.unity.com/packages/vfx/shaders/subsurface-scattering-124637

He said it does not do double sided geo

**Comment by justinjacot** [comment]
As a side note, the SSS shader that comes with the amplify shader editor is good looking, but as far as I can tell doesn't work well with shadows, so that's out.  I emailed them and they confirmed the limitation.  

I followed this tutorial (https://www.youtube.com/watch?v=RPkhbjmmvS0).  It looked good except the SSS effect has a hard cutoff with the light radius.  It does support double sided geo though, so I would like to experiment more with it.

This shows how good the new HDRP version looks (https://www.youtube.com/watch?v=5ZymuPouxxc&t=19s).  Quite good however, in my experimentation with it, on simple shapes like cubes (or all of the models in our game), there is a visible quality stepping, just like the volumetric lighting that they are building in.  I would hope and expect both features to improve as HDRP matures, and so ultimately, it may be a good solution as it does do double sided geo as well.

**Comment by justinjacot** [comment]
If you read the reviews on this one (https://assetstore.unity.com/packages/vfx/shaders/subsurface-scattering-shader-3254), it does not really support shadows either.  It also only works in forward rendering, and even then, only directional lights work, unless you hack it

**Comment by justinjacot** [comment]
This one is cheap and simple and runs on web, so likely will work for us.
https://assetstore.unity.com/packages/2d/textures-materials/translucent-scatter-99826

There's a web demo

**Comment by justinjacot** [comment]
This one has potential as it seems fairly newly updated, and claims to work with all light types
https://assetstore.unity.com/packages/vfx/shaders/pbs-subsurface-scattering-100332

Cheap as well, but not much info to go on.  No demo from what I can tell

**Comment by GameSMRT** [comment]
Shaders have nothing to do with how Unity is drawing shadows unless they are emitting emissive light or set to be unlit
(Ignore lights) So I’m not sure what exact artifact you are seeing/they are talking about that makes it “not work” with shadows.

SSS is basically a fake effect the shader is doing. Grabbing info from the scene like the light and camera angle to cause the color of the light behind it to show through.
When the light is in front of the cranes in my video. My cranes look exactly as they would with non-SSS shader as they should. 
Lux doesn’t affect the shadows at all in my scene.

Instead of looking at these shaders that seem to be purposed for humans or wax like situations...
https://assetstore.unity.com/packages/vfx/shaders/advanced-foliage-shaders-v-5-68907
This is a shader I’ve used before that has SSS and specifically supports card geo because that’s what foliage is. And I know for sure these aren’t unlit shaders so they aren’t going to mess with shadows.

**Comment by justinjacot** [comment]
The shadow issue is that you cannot turn shadows on on your light.  You cannot cast shadows onto it...so its useless
I was just looking at that shader.  It might make a lot of sense.  Do you have it already?  It;'s by the same lux guys

**Comment by GameSMRT** [comment]
Ahh that makes way more sense now. They made the SSS effect but didn’t enable the shader to be lit by the engine. 

I have an old version of the system because I asked Kelvin to buy it and it’s in old projects I still have. I’ll dig it up. We’d have to buy it again though to get the latest version because kelvin bought it on his account.

**Comment by justinjacot** [comment]
Before we buy anything more, I wanna see what we can do with the amplify shader and I would want to know what shadows look like coming through.  The paper look you achieved in that demo is really good, and it has a lot to do with the shadows that are coming through.  I'm not sure any of the other SSS shaders would do that (which I think is what you were getting at).  Is there any control over how much the shadow blurs behind?  Is it based on "thickness"?

**Comment by justinjacot** [comment]
It may almost be worth getting the LUX thing working that you already have set up.  Do we know what the issue is?

**Comment by GameSMRT** [comment]
As I suspected the secret to pull off the effect is a translucency map.

All SSS shaders by default will just “blow” the light thought the entire model. The shader asks a question. With the current camera angle is there a light behind the model I’m shading? If so take the color of the light and color the model that color.
This question does not take into account any information on the models thickness.

To achieve the effect of only the thin parts of the model bleeding light though & the look of layered folds showing through. I baked a translucency map in xnormal but had to edit it to dial in the exact areas I wanted light to bleed through. And specifically made a two layer “fold” on the 3D cranes wing as seen in the ref image.

An AO map was also nessasary because it assisted in masking off the bleeding so thick areas don’t change.

**Comment by justinjacot** [comment]
Right.  Mkay.  Since you know more about it, I will leave you to it mostly, but if I come across anything, I will let you know.  Hdrp is still the only full option I have found

**Comment by GameSMRT** [comment]
Foilage shaders were a bust.
https://youtu.be/hgx5zBCoDrQ

- No ability to tint color the diffuse map
- No AO map slot.
- SSS is a color you pick, not based on the light color coming through.
- kinda is working but lux is way better.

I reimported the crane scene into a new project and verified it's still not working correctly.

Alright rather than wasting time buying and hoping the other SSS shaders work. Let's just get the paid version of since we know it works for both geo types and looks good., it's only $20
https://assetstore.unity.com/packages/vfx/shaders/lux-plus-physically-based-shader-framework-74897
says right in the description it's compatible with Unity 2018, it's the free version they stopped updating.

**Comment by justinjacot** [comment]
I'm into it.  Let's do it.  I can grab it on my account so it's all in the same place, or you can
